### PR TITLE
MM-33444 rest api GetRemoteClusterById

### PR DESCRIPTION
**Pull request by wiggin77**
#### Summary
This PR adds a new Rest API for fetching details for a remote cluster based on id.

`/api/v4/sharedchannels/getremote/{remote_id}`    GET

Returns JSON:
{
    display_name: string,
    create_at: int,
    last_ping_at: int
}

The user making the call must be a member of at least one channel shared with the remote.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-33444

API Docs:  https://github.com/mattermost/mattermost-api-reference/pull/619

#### Release Note
```release-note
NONE
```

### CODE REVIEW

File: api4/shared_channel.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+	}
+
+	c.RequireRemoteId()
+	if c.Err != nil {
+		return
+	}
+
+	// GetRemoteClusterForUser will only return a remote if the user is a member of at
+	// least one channel shared by the remote. All other cases return error.
+	rc, err := c.App.GetRemoteClusterForUser(c.Params.RemoteId, c.App.Session().UserId)
```
Comment (nronas) [inline api4/shared_channel.go:29]: This call can return both an error due to noRowsFound but also a "canonical" error. What's the reason that we are returning 404 always here? I would have expected a 404 on the former case and a 500 on the latter.

**Review by nronas — APPROVED** [review]
Only a clarification, functionality-wise this LGTM

File: api4/shared_channel.go (hunk 1 above)
Comment (wiggin77) [reply api4/shared_channel.go]: Main reason was I was not using AppError in the app layer.  AppError is the biggest source of crashes in the entire app and I wish we'd get rid of it.  Anyway, I made this API use AppError and it now differentiates between 500 and 404.  Fixed via https://github.com/mattermost/mattermost-server/pull/17063/commits/9288e226865aae981b33cc140fe383c4028505c8

**Review by nronas — APPROVED** [review]

File: services/remotecluster/service.go [hunk 2]
```diff
@@ -49,6 +50,23 @@ type ServerIface interface {
 	GetMetrics() einterfaces.MetricsInterface
 }
 
+// RemoteClusterServiceIFace is used to allow mocking where a remote cluster service is used (for testing).
+// Unfortunately it lives here because the shared channel service, app layer, and server interface all need it.
+// Putting it in app layer means shared channel service must import app package.
```
Comment (isacikgoz) [inline services/remotecluster/service.go:55]: Could this still be added to `app` package if some types could be defined in `model` package? It's a good practice to define interfaces where they are _used_.

File: app/shared_channel.go [hunk 3]
```diff
@@ -98,6 +99,20 @@ func (a *App) HasRemote(channelID string, remoteID string) (bool, error) {
...
 
+func (a *App) GetRemoteClusterForUser(remoteID string, userID string) (*model.RemoteCluster, *model.AppError) {
+	rc, err := a.Srv().Store.SharedChannel().GetRemoteForUser(remoteID, userID)
+	if err != nil {
+		var nfErr *store.ErrNotFound
+		switch {
+		case errors.As(err, &nfErr):
+			return nil, model.NewAppError("GetRemoteClusterForUser", "api.context.remote_id_invalid.app_error", nil, nfErr.Error(), http.StatusNotFound)
+		default:
+			return nil, model.NewAppError("GetRemoteClusterForUser", "api.context.remote_id_invalid.app_error", nil, err.Error(), http.StatusNotFound)
```
Comment (isacikgoz) [inline app/shared_channel.go:110]: Is this still a "not found" error? If so, there is no need for the `switch` block.

File: services/remotecluster/service.go [hunk 4]
```diff
@@ -127,6 +145,14 @@ func (rcs *Service) Shutdown() error {
 	return nil
 }
 
+// Active returns true if this instance of the remote cluster service is active.
+// The active instance is responsible for pinging and sending messages to remotes.
+func (rcs *Service) Active() bool {
+	rcs.mux.Lock()
```
Comment (isacikgoz) [inline services/remotecluster/service.go:151]: Is it intentional to share same mutex for methods like `Add/Remove TopicListener`?

**Review by isacikgoz — COMMENTED** [review]
A lot of effort here 💪 Have some questions regarding to implementation and some suggestions to make API a bit more idiomatic.

File: services/remotecluster/service.go (hunk 2 above)
Comment (wiggin77) [reply services/remotecluster/service.go:55]: I agree that is a good practise, however `model` is not where the interface is used either.  Given a choice between putting the interface where the implementation is versus a place where it is neither used nor implemented, I picked where it was implemented.

File: app/shared_channel.go (hunk 3 above)
Comment (wiggin77) [reply app/shared_channel.go:110]: I understand a single case + default switch might look strange, but that is the defacto standard for our app.  That exact switch occurs in 70 places in the app, including new code.

File: services/remotecluster/service.go (hunk 4 above)
Comment (wiggin77) [reply services/remotecluster/service.go:151]: Yes.  There are low contention (rarely called) APIs and don't need another mutex.

File: web/context.go [hunk 5]
```diff
@@ -271,6 +275,11 @@ func NewInvalidRemoteClusterTokenError() *model.AppError {
 	return err
 }
 
+func NewJSONEncodingError() *model.AppError {
+	err := model.NewAppError("Context", "api.context.json_encoding.app_error", nil, "", http.StatusUnauthorized)
```
[the remark this answers is not in the document]
Comment (wiggin77) [reply web/context.go]: Fixed via https://github.com/mattermost/mattermost-server/pull/17063/commits/fd8999c0d7b36cdba0733d2fadbf9ca23708b71a

File: api4/shared_channel.go [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+	"net/http"
+)
+
+func (api *API) InitSharedChannels() {
+	api.BaseRoutes.SharedChannels.Handle("/getremoteinfo/{remote_id:[A-Za-z0-9]+}", api.ApiSessionRequired(getRemoteClusterById)).Methods("GET")
+}
+
+func getRemoteClusterById(c *Context, w http.ResponseWriter, r *http.Request) {
+	// make sure remote cluster service is enabled.
+	if _, appErr := c.App.GetRemoteClusterService(); appErr != nil {
```
[the remark this answers is not in the document]
Comment (wiggin77) [reply api4/shared_channel.go]: Fixed via https://github.com/mattermost/mattermost-server/pull/17063/commits/fd8999c0d7b36cdba0733d2fadbf9ca23708b71a

File: api4/shared_channel.go [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+		c.Err = appErr
+		return
+	}
+
+	c.RequireRemoteId()
+	if c.Err != nil {
+		return
+	}
```
[the remark this answers is not in the document]
Comment (wiggin77) [reply api4/shared_channel.go]: Fixed via https://github.com/mattermost/mattermost-server/pull/17063/commits/fd8999c0d7b36cdba0733d2fadbf9ca23708b71a

File: api4/shared_channel.go [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+package api4
+
+import (
+	"encoding/json"
+	"net/http"
+)
+
+func (api *API) InitSharedChannels() {
+	api.BaseRoutes.SharedChannels.Handle("/getremoteinfo/{remote_id:[A-Za-z0-9]+}", api.ApiSessionRequired(getRemoteClusterById)).Methods("GET")
```
[the remark this answers is not in the document]
Comment (wiggin77) [reply api4/shared_channel.go]: Fixed via https://github.com/mattermost/mattermost-server/pull/17063/commits/fd8999c0d7b36cdba0733d2fadbf9ca23708b71a

File: services/remotecluster/service.go (hunk 2 above)
Comment (isacikgoz) [reply services/remotecluster/service.go:55]: Sorry I should've be more clear about this, I meant defining the interface in the `app` package but defining shared bits in the `model` package such as function types, structs etc.

File: app/shared_channel.go (hunk 3 above)
Comment (wiggin77) [reply app/shared_channel.go]: Check latest code, I believe the second case does return internal error.

File: services/remotecluster/service.go (hunk 2 above)
Comment (wiggin77) [reply services/remotecluster/service.go:55]: The interface is needed by shared channel service and app.  If defined in app then I need to import app into shared channel service which is not wanted.  Shared channel service uses a server interface that includes an API returning this interface, which is why we can't simply have two definitions of that interface.  It don't like it either, but it is an artifact of how server owns service instances and how we can mock server and those instances.  BTW, there is precedent for breaking the interface where it is used guideline in the standard lib.

Comment (isacikgoz) [reply services/remotecluster/service.go:55]: I see, thanks for the clarification.

**Review by isacikgoz — APPROVED** [review]
LGTM

**Pull request merged**